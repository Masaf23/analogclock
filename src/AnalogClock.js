import React, { Component } from 'react';
import Util from './Util';
import { ClockContainer, ClockBaseBorder, ClockBase, ClockCenter } from './ClockComponents';
import Hand from './Hand.js';

class AnalogClock extends Component {

    constructor(props) {
        super(props);
        this.state = this.initClock();
    }

    initClock() {
        const date = new Date();
        return {
            seconds: date.getSeconds(),
            minutes: date.getMinutes(),
            hours: Util.getHourIn12HrFormat(date.getHours())
        }
    }

    setupTime() {
        const date = new Date();
        this.setState({
            seconds: date.getSeconds(),
            minutes: date.getMinutes(),
            hours: Util.getHourIn12HrFormat(date.getHours())
        })
    }

    setupInterval() {
        this.interval = setInterval(() => this.setupTime(), 1000);
    }

    componentDidMount() {
        this.setupInterval();
    }

    componentDidUpdate(prevProps) {
        if (prevProps.useCustomTime !== this.props.useCustomTime) {
            if (this.props.useCustomTime) {
                clearInterval(this.interval);
                this.setState({
                    seconds: undefined,
                    minutes: undefined,
                    hours: undefined
                });
            } else {
                this.setupInterval();
            }
        }
    }

    render() {
        const { width, border, borderColor, baseColor, centerColor, centerBorderColor, handColors } = this.props;
        const { seconds, minutes, hours } = this.props.useCustomTime ? this.props : this.state;
        return (
            <ClockContainer width={width}>
                <ClockBaseBorder border={border} borderColor={borderColor}>
                    <ClockBase baseColor={baseColor}>
                        <ClockCenter centerColor={centerColor} centerBorderColor={centerBorderColor} />
                        <Hand type="second" seconds={seconds} handColors={handColors} />
                        <Hand type="minute" minutes={minutes} handColors={handColors} />
                        <Hand type="hour" hours={hours} minutes={minutes} handColors={handColors} />
                    </ClockBase>
                </ClockBaseBorder>
            </ClockContainer>

        )
    }

    componentWillUnmount() {
        clearInterval(this.interval);
    }
}

export default AnalogClock;