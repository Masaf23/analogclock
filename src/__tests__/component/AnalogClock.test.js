import React from "react";
import "jest-styled-components";
import { shallow } from "enzyme";
import "../../setupTests";

import AnalogClock from "../../AnalogClock.js";

describe("AnalogClock component", () => {
  const RealDate = Date;

  afterEach(() => {
    global.Date = RealDate;
  });

  it("matches snapshot", () => {
    const constantDate = new Date("2021-10-10T10:10:10");

    Date = class extends Date {
      constructor() {
        return constantDate;
      }
    };

    let options = {
      width: "300px",
      border: true,
      borderColor: "#2e2e2e",
      baseColor: "#17a2b8",
      centerColor: "#459cff",
      centerBorderColor: "#fff",
      handColors: {
        second: "#d81c7a",
        minute: "#fff",
        hour: "#fff",
      },
    };
    const wrapper = shallow(<AnalogClock {...options} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("matches snapshot when custom time is passed", () => {
    let options = {
      useCustomTime: true,
      seconds: 22,
      minutes: 22,
      hours: 22,
      width: "300px",
      border: true,
      borderColor: "#2e2e2e",
      baseColor: "#17a2b8",
      centerColor: "#459cff",
      centerBorderColor: "#fff",
      handColors: {
        second: "#d81c7a",
        minute: "#fff",
        hour: "#fff",
      },
    };
    const wrapper = shallow(<AnalogClock {...options} />);
    expect(wrapper).toMatchSnapshot();
  });
});
