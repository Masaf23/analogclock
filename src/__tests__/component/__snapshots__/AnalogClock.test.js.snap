// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnalogClock component matches snapshot 1`] = `
<styled.div
  width="300px"
>
  <styled.div
    border={true}
    borderColor="#2e2e2e"
  >
    <styled.div
      baseColor="#17a2b8"
    >
      <styled.div
        centerBorderColor="#fff"
        centerColor="#459cff"
      />
      <Hand
        handColors={
          Object {
            "hour": "#fff",
            "minute": "#fff",
            "second": "#d81c7a",
          }
        }
        seconds={10}
        type="second"
      />
      <Hand
        handColors={
          Object {
            "hour": "#fff",
            "minute": "#fff",
            "second": "#d81c7a",
          }
        }
        minutes={10}
        type="minute"
      />
      <Hand
        handColors={
          Object {
            "hour": "#fff",
            "minute": "#fff",
            "second": "#d81c7a",
          }
        }
        hours={10}
        minutes={10}
        type="hour"
      />
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`AnalogClock component matches snapshot when custom time is passed 1`] = `
<styled.div
  width="300px"
>
  <styled.div
    border={true}
    borderColor="#2e2e2e"
  >
    <styled.div
      baseColor="#17a2b8"
    >
      <styled.div
        centerBorderColor="#fff"
        centerColor="#459cff"
      />
      <Hand
        handColors={
          Object {
            "hour": "#fff",
            "minute": "#fff",
            "second": "#d81c7a",
          }
        }
        seconds={22}
        type="second"
      />
      <Hand
        handColors={
          Object {
            "hour": "#fff",
            "minute": "#fff",
            "second": "#d81c7a",
          }
        }
        minutes={22}
        type="minute"
      />
      <Hand
        handColors={
          Object {
            "hour": "#fff",
            "minute": "#fff",
            "second": "#d81c7a",
          }
        }
        hours={22}
        minutes={22}
        type="hour"
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
